modules/message-lane/src/lib.rs

```diff
@@ -335,22 +335,24 @@ decl_module! {
 
 				// reward relayers that have delivered messages
 				// this loop is bounded by `T::MaxUnconfirmedMessagesAtInboundLane` on the bridged chain
-				for (nonce, relayer) in lane_data.relayers {
-					// we may receive the same confirmation multiple times
-					if nonce < received_range.0 {
-						continue;
-					}
-
-					let message_data = OutboundMessages::<T, I>::get(MessageKey {
-						lane_id,
-						nonce,
-					}).expect("message was just confirmed; we never prune unconfirmed messages; qed");
+				for (nonce_low, nonce_high, relayer) in lane_data.relayers {
+					for nonce in nonce_low..nonce_high + 1 {
+						// we may receive the same confirmation multiple times
+						if nonce_high < received_range.0 {
```

Comment:
1) I think you wanted to compare `nonce`, not `nonce_high` here, right?
2) let's move before the inner loop instead. Something like that:
```rust
let nonces_begin = sp_std::cmp::max(nonce_low, received_range.0);
let nonces_end = sp_std::cmp::min(nonce_high, received_range.1);
for nonce in nonces_begin..=nonces_end {
 ...
}
```
The second check (`nonces_end`) isn't absolutely necessary iirc - just in case...

Reply:
This is really good. I updated the code accordingly. Thanks for spotting the mistake with `if nonce_high < received_range.0` - you're absolutely right it was supposed to be `nonce`. What it should be is actually a check 
```
if nonce_low > nonce_end {
 continue;
}
```
to avoid rewarding the relayers that dispatched messages ahead of the received range. The inverse case for received ranges from before the dispatched messages range are already discarded by `lane.confirm_delivery()`.